Sources/Algorithms/Product.swift

```diff
@@ -183,10 +193,219 @@ extension Product2: Collection where Base1: Collection {
+    return index
+  }
+  
+  @usableFromInline
+  internal func offsetBackward(_ i: Index, by distance: Int) -> Index {
+    guard let index = offsetBackward(i, by: distance, limitedBy: startIndex)
+      else { fatalError("Index is out of bounds") }
+    return index
+  }
+  
+  @usableFromInline
+  internal func offsetForward(
+    _ i: Index, by distance: Int, limitedBy limit: Index
+  ) -> Index? {
+    assert(distance >= 0)
+    assert(limit >= i)
+    
+    if limit.i1 == i.i1 {
+      // Delegate to `base2` if the offset is limited to `i.i1`.
+      //
+      //             i.i2      limit.i2
+      //              v           v
+      // i.i1 > [x x x|x x x x x x|x x x]
+      
+      return base2.index(i.i2, offsetBy: distance, limitedBy: limit.i2)
+        .map { i2 in Index(i1: i.i1, i2: i2) }
+    }
+    
+    
+    if let i2 = base2.index(i.i2, offsetBy: distance, limitedBy: base2.endIndex) {
+      // `distance` does not overflow `base2[i.i2...]`.
+      //
+      //             i.i2         i2
+      //              v           v
+      // i.i1 > [x x x|x x x x x x|x x x]
+      //        [     |> > > > > >|     ]   (`distance`)
+      
+      return convertIndex(i.i1, i2)
+    }
+    
+    let suffixCount = base2[i.i2...].count
+    let remaining = distance - suffixCount
+    let nextI1 = base1.index(after: i.i1)
+    
+    if limit.i1 == nextI1 {
+      // Delegate to `base2` if the offset is limited to `nextI1`.
+      //
+      //               i.i2
+      //                v
+      //   i.i1 > [x x x|x x x x x x x x x]
+      // nextI1 > [x x x x x x x x x|x x x]
+      //                            ^
+      //                         limit.i2
+      
+      return base2.index(base2.startIndex, offsetBy: remaining, limitedBy: limit.i2)
+        .map { i2 in Index(i1: nextI1, i2: i2) }
+    }
+    
+    if let i2 = base2.index(base2.startIndex, offsetBy: remaining, limitedBy: i.i2) {
+      // `remaining` does not overflow `base2[..<i.i2]`.
+      //
+      //                           i.i2
+      //                            v
+      //   i.i1 > [x x x x x x x x x|x x x]
+      //          [                 |> > >]   (`suffixCount`)
+      //          [> > >|                 ]   (`remaining`)
+      // nextI1 > [x x x|x x x x x x x x x]
+      //                ^
+      //                i2
+      
+      return Index(i1: nextI1, i2: i2)
+    }
+    
+    let prefixCount = base2[..<i.i2].count
+    let base2Count = prefixCount + suffixCount
+    let base1Distance = remaining / base2Count
+    
+    guard let i1 = base1.index(nextI1, offsetBy: base1Distance, limitedBy: limit.i1)
+      else { return nil }
+    
+    // The distance from `base2.startIndex` to the target.
+    let base2Distance = remaining % base2Count
+    
+    let base2Limit = limit.i1 == i1 ? limit.i2 : base2.endIndex
+    return base2.index(base2.startIndex, offsetBy: base2Distance, limitedBy: base2Limit)
+      .map { i2 in Index(i1: i1, i2: i2) }
+  }
```

Comment:
I think my main question with this patch is how much do these special cases in `offsetForward` and `offsetBackward` buy us? It's a lot to hold in your head.

Can we get away with just the main "loop"?

```swift
internal func offsetForward(
  _ i: Index, by distance: Int, limitedBy limit: Index
) -> Index? {
  assert(distance >= 0)
  assert(limit >= i)

  let suffixCount = base2[i.i2...].count
  let remaining = distance - suffixCount
  let nextI1 = base1.index(after: i.i1)
  let prefixCount = base2[..<i.i2].count
  let base2Count = prefixCount + suffixCount
  let base1Distance = remaining / base2Count

  guard let i1 = base1.index(nextI1, offsetBy: base1Distance, limitedBy: limit.i1)
    else { return nil }

  // The distance from `base2.startIndex` to the target.
  let base2Distance = remaining % base2Count

  let base2Limit = limit.i1 == i1 ? limit.i2 : base2.endIndex
  return base2.index(base2.startIndex, offsetBy: base2Distance, limitedBy: base2Limit)
    .map { i2 in Index(i1: i1, i2: i2) }
}
```